### DISCUSSION TITLE
[Feature]: Estimate Location from other photos

### DISCUSSION THREAD

**Discussion by r3nor** [immich-app/immich#1675] [Feature Request]
### Feature detail

So Photoprism has a feature that given a 24h time span, it sets the geotag of the images from that span with the geolocation of other images that were taken the same day.

I think this is a very useful feature, but 24h I think is too much and can be innacurate. 

It would be a good feature if Immich could get Geolocation estimations from close-taken pictures for those that do not have Geolocation metadata in short timespans.

### Platform

Server

**Comment by alextran1502** [comment]
Hmm, since most of the picture uploaded from the phone already has geolocation in the image's EXIF. I am not sure if we need this info at all

**Reply by mirisbowring** [reply]
In my case, My Camera only adds GEO Tags when connected to my Phone via Bluetooth. Sometimes (on a party) I am to far away from my phone or i started the camera and shot some photos before the bluetooth connection came up. Or I just forget to open the geo app on my phone

So assuming a timeframe of around 10-60 minutes, the location should still be pretty accurat (better than nothing) and could improve such user experiance :)

**Reply by blublub** [reply]
> Hmm, since most of the picture uploaded from the phone already has geolocation in the image's EXIF. I am not sure if we need this info at all

Yes and no. Many ppl still use "normal cameras" for some photos - this would be a real cool addition to add geo-tags to "non phone images".
Some modern cameras have GPS or get it via bluetooth fromt the phone but for the latter my expirience is that it is buggy

**Reply by mpschr** [reply]
pictures shared trough messaging apps (think famili/friends gathering) often/normallly have stripped geo-location

**Comment by r3nor** [comment]
Yes but even with mobile not all photos may have geotags. I explain why I thought about this:

I went on vacation. I made some exterior photos to my family, but then we went to visit a cave where there was no GPS signal, the photos I took in there have no geotags. Also, later that same day we went to eat to a restaurant where there wasn't GPS signal too.

This is not a rare situation. In my case, I don't have any GPS signal on almost any interior spaces...

**Comment by EnderPicture** [comment]
To add, Google photos have this feature, though it also uses location history data. It would be helpful for phone + camera setups where a traveller would use phones to take snapshots while dedicated cameras with no GPS for more artistic/composed landscapes.

**Comment by AngelaDMerkel** [comment]
I know this feature request is quite old, but it would be very useful to import the takeout location data from Google Timeline, or otherwise to be able to import gpx or geoJSON. It's possible to write to the files before adding to Immich, but considering this application is pitched as a fairly direct competitor to Google Photos, it might be useful to be able to pull from Google's own collection of our data.

**Reply by maxshcherbina** [reply]
I would post this as a separate feature request if you haven't already. I wonder if anyone has made a python script or something that can extract the metadata and write it back into the photo. When I have some free time I'll investigate

**Reply by AngelaDMerkel** [reply]
exiftool can read from gpx and write which makes this pretty easy. There are a few python scripts with the same functionality but I don't recall which I've used in the past. The problem with gpx is the fuzziness where GPS and image times don't line up perfectly and there isn't any intelligent matching. That takes a little bit more work

**Comment by ShinasShaji** [comment]
I would like to mention a situation where, let's say, you were to save images from the web or take a couple screenshots at around the same time as shooting some photos with geolocation information. This would add the location data to the saved web images and screenshots as well, and I have had this happen during the time I used PhotoPrism. 

Side note, PhotoPrism would also stack such screenshots and saved images due to them being around the same time and location, making the problem further annoying. I have had to turn off Location Estimates and Stack by location and time behaviors to make this aspect of my library usable.

I think this would be a significant issue considering Immich is also meant to back up your phone's photo library, which will invariably have such saved images and screenshots.

**Comment by dominikheinz** [comment]
@r3nor 
How exactly is this location approximation feature supposed to work?
Say I take 3 pictures on the same day, the first is in city A and the other 2 are in city B. Both cities could be hundreds of kilometers apart.
Now imagine we had GPS/Location data automatically added for picture 1 and 3. How are we supposed to approximate the location of picture 2? We can't just select a middle point between the two spots.

I am very interested in a feature like this, but I can't see how this would effectively work from a technical standpoint.
Currently, I am in the process of going through old vacation pictures, which had no GPS information on them. And then I manually tag them, utilizing Google Maps Street View, Research of the areas, checking addresses of guesthouses and so on... It is a very tedious (manual) process. How would you automate/approximate this?

**Reply by jrasm91** [reply]
I think if they were taken between other photos (timewise) by the same phone by the same person you could copy the gps coordinates from the ones that did have them.

**Reply by dominikheinz** [reply]
Hmm, I feel like this would be extremely unreliable. Pictures could be taken only a few minutes apart from each other, but the location could be completely different. (Think about taking pictures out of a fast moving vehicle for example).

**Reply by jrasm91** [reply]
Right, but I think the argument is that having it have coordinates in the same city would be better than it not having any coordinates at all and subsequently not showing up in the map view, no data in the info panel, etc.

**Reply by AngelaDMerkel** [reply]
The way this is handled when manually tagging images is to establish a duration of time and distance from (usually margin of error) a know GPS position. If there is GPS location data every 20 meters, if might also be reasonably to estimate the location in a line between two of the points. 
In other words, if you took a photo of the London Eye with your phone and there was know GPS data, then you might allow that any image taken without two minutes of that image is likely to be the same location. If you then cross the bridge and take and image at Westminster, you might assume that the location of a photo four minutes from then is in some location in between the two. Obviously the result would likely be an image in the Thames, but that's a pretty basic yet useful implementation.

**Reply by AngelaDMerkel** [reply]
@r3nor @jrasm91 It appears exif has two fields describing margin of error: `GPSDOP` and `GPSHPositioningError`. The former is perfect in these instances where results are fuzzy. Otherwise only putting in city info rather than precise location would be workable

**Comment by kskoba** [comment]
There’s a good example of an implementation of this type of function in the OSX Photos CLI software found here on GitHub. Would be useful when importing a mix of DSLR photos and phone photos for sure. very niche cases (taking a photo from a moving vehicle where GPS data is enabled for one photo but not the next photo 2 minutes later) can be addressed/caught with some of the settings. 

https://rhettbull.github.io/osxphotos/cli.html

**Comment by scblakely** [comment]
I would like this feature, as well. I take lots of photos with my non-phone camera (and no GPS), and my wife is generally taking a few snaps with her phone. Being able to co-locate (based on time) un-tagged photos with the geo-tagged images would be fantastic.

Set a time resolution for the association, and suddenly un-tagged images get located (with some uncertainty).

**Comment by GregTroar** [comment]
Also very interested in this feature.
Currently on vacation taking pictures with the phone and I deactivate gps when not taking pictures to limit battery draw. But sometimes forget to turn back on gpd for a few pictures.
Would be really nice.
I was thinking in another way a bit like merging two faces into one.
You select pictures without geotag and then select a picture with the geotag which was taken at same location and it would apply the location of the picture having 5he geo location to all the other ones you previously selected. That would allow to have 3xact same location rather than an 3etinated one by selecting on the map.

**Comment by lunae-f** [comment]
I am asking for this feature. I usually take pictures with my cell phone, but when I travel I use my still camera (Olympus E-PL5). Of course, I can create albums based on date, but the ability to automatically geo-tag would be more useful.

**Comment by semiligneous** [comment]
Would also really like this feature. Trying Immich, but currently a PhotoPrism user, and when comparing both service's map/location features it really stands out on my library. I have tens of thousands of photos from "normal" cameras (DSLR etc) which don't have any exif GPS data. However in PhotoPrism's map I see almost all of these in approximately the right location.

Of course they are not exact, but they are in the right region, usually within a few km, and that's essentially what the map view is for - to be able to browse from the world view down, and see/remember what you did in various places. The location not being precise is, to me, far more preferable than missing out on seeing all those photos at all!

Regarding comment about screenshots getting geo-tagged in PhotoPrism due to this method, in my setup screenshots get categorised as "low quality" and don't get shown on the map or timeline views, which seems like a good way to do it. Having a toggle for "show/hide screenshots" would be another good option if Immich wanted to think about that. Or just, don't estimate geo location for screenshots.

This was/is probably the one biggest feature that might make PhotoPrism my preferred tool.

Thanks!

**Reply by AngelaDMerkel** [reply]
It's possible to have less precise GPS data where it isn't available and I think that is a fine compromise, as you say. I would really like it to even just tell me the city or rough region. This is more helpful for hiking especially if it's years prior where you might not recall a multiday route.

**Comment by jforseth210** [comment]
The big concern I see with this is that photos from multiple different people could get garbled really easily here. For instance, I was part of a shared (non-Immich) photo album with some friends a while back. I then downloaded these files and loaded them into Immich. Friend A might be at place X, Friend B at place Y taking photos at the same time. 

Suppose A used geotagging, and B didn't. How do we make sure that B's photos don't show up at X?

**Reply by blublub** [reply]
> The big concern I see with this is that photos from multiple different people could get garbled really easily here. For instance, I was part of a shared (non-Immich) photo album with some friends a while back. I then downloaded these files and loaded them into Immich. Friend A might be at place X, Friend B at place Y taking photos at the same time.
> 
> Suppose A used geotagging, and B didn't. How do we make sure that B's photos don't show up at X?

Valid concern. 

This could be prevented with that not being an automated feature, more a guided one:

1. specify time / range
2. Immich lists cameras without and with GPS
3.  User selects from which camera a GPS location should be transferred and ho close the timestamp has to be

**Comment by semiligneous** [comment]
Yeah that’s valid and something I did think about, and would need to be addressed somehow.

I guess firstly, it’s probably an opt in feature. If you don’t think it will work for your setup you could not enable the ‘estimate locations’ feature and there will certainly be some people that this will be better for. But perhaps you want it for some folders/events/scenarios but not others due to the cameras physically being in different locations at the same / similar time.

Photoprism manages it somehow; whether it causes strange effects/impact on a  user’s library would of course be specific to each user. For my library/ies it works very well in practice. But I don’t have whole albums/folders of other people’s photos. Maybe a few photos every now and again.

Do you have the opportunity to run photoprism on the library which you have this concern to see what happens in practice?


One way would perhaps be to expose the ‘link’ that is used to make the estimate as a metadata field/property of the photo. And then allow via the UI to ‘unlink’ from an incorrect estimate or even ‘relink’ to a note appropriate estimate.

For example:


Photo 1 is taken by an iPhone with accurate GPS at 09:05:00

Photo 2 is taken with a normal camera with no location data at 09:15:00, the photo is taken by the owner of the iPhone a few hundred meters away from Photo 1.

Photo 3 is taken by another iPhone but by different person at 09:10:00 in another city, with geolocation enabled.


To be accurate, you want Photo 2’s location to be estimated by using Photo 1’s location. However, since the timestamps are closer to Photo 3 it will probably get linked to that one, leading to an inaccurate map display.


If you open Photo 2 and go to meta data > estimated location > there could be a list of photos that were used. You could deselect one; ‘don’t use this photo to estimate location’, or even ‘don’t use photos taken by this camera to estimate the location’ or even further ‘don’t use this camera to estimate locations for ANY photos in this Immich instance’.

Lots of options to think about! Not trivial understand…




----- Original message -----

**Reply by AngelaDMerkel** [reply]
I think the solution is easier than this. Only the user uploading the images is used to refer to their other uploaded images. This can be true for every user. I am not sure if immich has this robust of a user tracking system though

**Reply by jforseth210** [reply]
Tested Photoprism's implementation. Found a few false results, but was overall impressed with how well it worked and can see the appeal of the feature a lot more now.

Still, I don't like the idea of Immich changing the metadata of my pics based on a guess. I think this would be a good feature to add the utilities section, with duplicate detect. Have it ask, "Was this photo taken here?" and give you a little pin you can drag around.

**Comment by jforseth210** [comment]
> Do you have the opportunity to run photoprism on the library which you have this concern to see what happens in practice?

Sure! I have a Photoprism instance I was playing around with not to long ago. Hopefully this weekend I'll have some time to spin it back up, get those photos imported, and see what it does. Will report back with my findings.

**Reply by AngelaDMerkel** [reply]
Some time ago when I tested, the results were adequate for me.

**Reply by Dieghito72** [reply]
I recently switched from Photoprism to Immich and I have to say is the only feature I'm missing (for all the rest Immich is superior from my point of view) since most of the photo are shot from reflex camera with no GPS info.
The algorithm used by Photoprism is not perfect (especially if move from one place to another in the same day), but the result was more than acceptable.

**Comment by semiligneous** [comment]
Any estimated locations should certainly only be stored as part of the separate metadata/index - not by editing the exif info on the files themselves. This is how Photoprism does it - all my libraries are mounted read/only to ensure this doesn't happen, and the features all work.

----- Original message -----

**Reply by jforseth210** [reply]
This is probably out of the scope of this discussion, but does this make migrating away from Immich/Photoprism harder?

**Comment by semiligneous** [comment]
Surely no more or less than any of the other features of either solution - such as face-detection or tags/labels? Those, and estimated locations, are all properties that are created and stored independent of the actual images. Moving to/from either platform inherently requires such information to be lost and then re-created in the new platform. Both can also be used side-by-side indefinitely if required.



----- Original message -----

**Reply by jforseth210** [reply]
I'm mainly coming from Digikam, which seems to try to do things in a portable way, including facial recognition and tagging. While this doesn't work super great in practice

I guess I never thought about whether Immich and Photoprism where doing the same.

**Reply by petersuba** [reply]
They may be stored independent from pictures NOW, but there are feature requests open for eventually moving such properties (also) to photo metadata directly (e.g. EXIF or sidecar), for example see #4348

**Reply by AngelaDMerkel** [reply]
I think that inferred data should never be written as metadata to originals. It's a useful tool for visualising where images were made, but not for actually locating a specific position in space. If users want to write these results to the originals, there is no reason there can't be such a tool, but it should be off by default.

**Reply by scblakely** [reply]
Certainly it should not be written to the Geolocation tag unless specifically requested, but it could (via Reverse Geocoding) fill in the IPTC location tags:

90:  City
92: Sub-location  
95: Province-State
100: Country-PrimaryLocationCode  
101: Country-PrimaryLocationName

**Reply by petersuba** [reply]
Yep my point is the request to read/write (i.e. sync) metadata across the DB and the image assets is a separate and valid feature request. I actually think that should have its own settings and tools.

**Comment by semiligneous** [comment]
Interesting concept - I was not aware that storing face data info to the images was a ‘thing’ but it makes sense and I can see how some people would want that. Personally if writing the face (or any other) metadata to the files got implemented in Immich I would probably choose to not use it for my use case, at least not yet. 

I think it would be something to enable/disable by each user, and also choose which type of data is OK to be written back to the images (facial, geo, geo-estimates etc), but I would still keep my "proper camera" images as external libraries mounted read only anyway so no estimated locations would be written. I upload those to storage location from SD card in other ways. 

I don't think this is directly impacting the "can Immich do location estimates" feature request though - but certainly something to take into account alongside other feature requests.

----- Original message -----

**Comment by matcpn** [comment]
Photoprism's implementation uses the most recent photo with a location taken before + after the one you're trying to estimate location on

Does a calculation how much you moved over how much time (velocity vector)

Looks at the timestamp of your photo, plops on a location estimate based on your velocity between the taking two photos 

Oversimplified example: user took photo 1 and photo 3, moved 10 m/s between the two geolocations, photo 2 was taken 10 seconds after photo 1, photo 2 must be 10 meters away from photo 1

source:
https://github.com/photoprism/photoprism/blob/develop/internal/entity/photo_estimate.go

**Reply by AngelaDMerkel** [reply]
I think this is a pretty good way to handle this problem if Immich won't add support for GPX track logs. Personally, I would be more than happy for an even simpler implementation where it just locates images within a certain time range from one another near each other. Otherwise you risk that it interpolates two photos taken between distant locations as being in the middle, as in a flight for example.

**Reply by jrasm91** [reply]
Thoughts on something like this? It isn't automatic but it seems like it might be a good manual implementation. Basically you look through your timeline and just copy/paste gps location data from one asset to another.

https://github.com/immich-app/immich/pull/20758

**Comment by semiligneous** [comment]
If that’s relatively simple to implement, it would be great to see in Immich. I agree it won’t be perfect, but in my opinion I’d prefer to see all my photos on a map even in a slightly incorrect position than not on the map at all (which is how it is today).

Some type of filter or icon or visual cue to show it’s an ‘estimated location’ rather than ‘gps location’ would be additionally useful/helpful.

Ability to switch the feature on or off, (maybe even per external library? In order to assist those who maybe have mixed sources, some of which may be more or less suited to this type of estimation) would be good.


----- Original message -----

**Comment by eharris** [comment]
I have been working on a standalone script to do this for my own photos collection, although it does it by re-writing the EXIF data directly on the photos.

My use case is my phone has GPS and my DSLR does not, and I want to use the GPS data from my phone's photos to populate the location tags for my DSLR photos.  An approximation is often good enough and much better than nothing.

For this to work well, it requires that the time on the phone and DSLR are well synced, which I always make sure of.
When I go to a new location (or move significantly), I try to always take a (sometimes throwaway) photo with my phone for the GPS data, and then take the vast majority of my photos with my higher quality DSLR.

My algorithm is to take the time and GPS location from the the four closest-in-time (2 before and 2 after) photos with GPS data, select which two to use, and do a linear interpolation/extrapolation of the location and tag it with those results.  I artificially change the GDOP values to reduce the precision of the location fix for these interpolated values based on the time difference between the two photos.

The reference photo selection process is a little complicated in an attempt to try to make the calculation more accurate.  I tend to prefer using the closest photos on either side (in time), but when one side has two much closer timestamps, it will extrapolate using the location and time of those instead of interpolating using the ones on either side.

There are some special circumstances:
- If there is no reference photo near enough in time, then GPS location is not added.
- If there is only one reference photo near enough in time, then the GPS location of that one is used as-is, and the GDOP is increased based solely on time distance from the reference.
- There is also a failsafe check, so that if there are nearby photos from multiple devices, it will not add location info since it can't be sure which device should be used or that all non-GPS-tagged photos were taken in the same location.
- I can override the failsafe by specifying which camera model (from the EXIF data) should be used as reference and which cameras are safe to add GPS data to.
- The overrides can also be done by filename patterns or photo resolution, in the case that the photo doesn't have usable model info in the tags, or there are more than one camera that identifies as the same model.

I also add another custom EXIF tag so that I can easily tell if the photo has real or interpolated GPS data, so that if I come up with better heuristics, I can go back and re-tag the photos.

Obviously these heuristics won't work on a whole-library basis, but they should work quite well within appropriately organized albums.

**Reply by eike2020** [reply]
Dear @eharris that sounds great. Do you mind sharing your script?

**Comment by midzelis** [comment]
This feature goes by the name of "geotagging" photos. And tools have been around for a long time that have done this. Search keywords: geotagging and interpolation. Even exiftool has a way to do this from gps track files: https://www.exiftool.org/geotag.html

That said, it would be nice to have this option in Immich. No other comments other than +1, this has my vote.

**Reply by AngelaDMerkel** [reply]
There is a really nice tool for this called [dawarich](https://github.com/Freika/dawarich) which even includes immich integration. I thought it was a feature I needed in immich, but having a standalone tool with good integration might be better.

**Comment by mpschr** [comment]
The most common case I see is pictures shared by groups through messaging apps that strip the exif data. This cases is also quite easy

**Comment by s-martin** [comment]
There's already a PR #20758 to implement this.
